process-migration-service/src/main/java/org/kie/processmigration/service/impl/MigrationServiceImpl.java

```diff
@@ -0,0 +1,278 @@
+
+    @Inject
+    private SchedulerService schedulerService;
+
+    @Override
+    public Migration get(Long id) throws MigrationNotFoundException {
+        TypedQuery<Migration> query = em.createNamedQuery("Migration.findById", Migration.class);
+        query.setParameter("id", id);
+        try {
+            return query.getSingleResult();
+        } catch (NoResultException e) {
+            throw new MigrationNotFoundException(id);
+        }
+    }
+
+    @Override
+    public List<MigrationReport> getResults(Long id) throws MigrationNotFoundException {
+        Migration m = get(id);
+        TypedQuery<MigrationReport> query = em.createNamedQuery("MigrationReport.findByMigrationId", MigrationReport.class);
+        query.setParameter("id", m.getId());
+        return query.getResultList();
+    }
+
+    @Override
+    public List<Migration> findAll() {
+        return em.createNamedQuery("Migration.findAll", Migration.class).getResultList();
+    }
+
+    @Override
+    @Transactional
+    public Migration submit(MigrationDefinition definition) throws InvalidMigrationException {
+        validateDefinition(definition);
+        Migration migration = new Migration(definition);
+        em.persist(migration);
+        if (ExecutionType.SYNC.equals(definition.getExecution().getType())) {
+            migrate(migration);
+        } else {
+            schedulerService.scheduleMigration(migration);
+        }
+        return migration;
+    }
+
+    @Override
+    @Transactional
+    public Migration delete(Long id) throws MigrationNotFoundException {
+        Migration migration = get(id);
+        em.remove(migration);
+        return migration;
+    }
+
+    @Override
+    @Transactional
+    public Migration update(Long id, MigrationDefinition definition) throws MigrationNotFoundException, ReScheduleException, InvalidMigrationException {
+        validateDefinition(definition);
+        Migration migration = get(id);
+        if (migration == null) {
+            return null;
+        }
+        if (!ExecutionStatus.SCHEDULED.equals(migration.getStatus())) {
+            throw new ReScheduleException("The migration is not scheduled and cannot be re-scheduled");
+        }
+        if (ExecutionType.SYNC.equals(definition.getExecution().getType())) {
+            throw new ReScheduleException("The migration execution type MUST be ASYNC");
+        }
+        migration.setDefinition(definition);
+        em.persist(migration);
+        schedulerService.reScheduleMigration(migration);
+        return migration;
+    }
+
+    @Override
+    @Transactional
+    public Migration migrate(Migration migration) throws InvalidMigrationException {
+        try {
+            Plan plan = planService.get(migration.getDefinition().getPlanId());
+            if (ExecutionStatus.CREATED.equals(migration.getStatus()) || ExecutionStatus.SCHEDULED.equals(migration.getStatus())) {
+                em.persist(migration.start());
```

Comment:
As discussed online, it won't

Reply:
By looking at it once again I think that this if statement is not needed at all. `migrate()` is called only when:
* Recovery is called - and then the migration is already managed since it comes straight from DB
* `@Timeout` occurred - in this case it is again already managed, since it is fetched from DB
* We migrate it right after `submit()` method is called which has, again, already fetched it from DB

So I think this if can be omitted completely.

Reply:
If the migration is recovered, the status is already `STARTED` and the `startedAt` field is set, if we omit this condition, the date will be reset.

Reply:
OK, but then it is sufficient to just call `migration.start()`.